sensors/movementsensor.go

```diff
@@ -0,0 +1,324 @@
+	ErrNoValidReadingObtained = errors.New("could not obtain a reading that satisfies the time tolerance requirement")
+)
+
+// TimedMovementSensor describes a sensor that reports the time the reading is from & whether or not it is
+// from a replay sensor.
+type TimedMovementSensor interface {
+	Name() string
+	DataFrequencyHz() int
+	TimedMovementSensorReading(ctx context.Context) (TimedMovementSensorReadingResponse, error)
+	Properties() MovementSensorProperties
+}
+
+// MovementSensorProperties contains information whether or not an IMU and/or odometer are supported.
+type MovementSensorProperties struct {
+	IMUSupported      bool
+	OdometerSupported bool
+}
+
+// TimedMovementSensorReadingResponse contains IMU and odometer sensor reading responses
+// as well as information about whether or not the readings are from a replay movement sensor.
+type TimedMovementSensorReadingResponse struct {
+	TimedIMUResponse      *TimedIMUReadingResponse
+	TimedOdometerResponse *TimedOdometerReadingResponse
+	IsReplaySensor        bool
+}
+
+// TimedIMUReadingResponse represents an IMU sensor reading with a time.
+type TimedIMUReadingResponse struct {
+	AngularVelocity    spatialmath.AngularVelocity
+	LinearAcceleration r3.Vector
+	ReadingTime        time.Time
+}
+
+// TimedOdometerReadingResponse represents an odometer sensor reading with a time.
+type TimedOdometerReadingResponse struct {
+	Position    *geo.Point
+	Orientation spatialmath.Orientation
+	ReadingTime time.Time
+}
+
+// MovementSensor represents a movement sensor.
+type MovementSensor struct {
+	name              string
+	dataFrequencyHz   int
+	imuSupported      bool
+	odometerSupported bool
+	replay            bool
+	sensor            movementsensor.MovementSensor
+}
+
+// Name returns the name of the movement sensor.
+func (ms *MovementSensor) Name() string {
+	return ms.name
+}
+
+// DataFrequencyHz returns the data rate in ms of the movement sensor.
+func (ms *MovementSensor) DataFrequencyHz() int {
+	return ms.dataFrequencyHz
+}
+
+// TimedMovementSensorReading returns data from the movement sensor and the time the reading is from & whether
+// it was a replay sensor or not.
+func (ms *MovementSensor) TimedMovementSensorReading(ctx context.Context) (TimedMovementSensorReadingResponse, error) {
+	var (
+		readingTimeAngularVel, readingTimeLinearAcc time.Time
+		readingTimePosition, readingTimeOrientation time.Time
+		angVel                                      spatialmath.AngularVelocity
+		linAcc                                      r3.Vector
+		position                                    *geo.Point
+		orientation                                 spatialmath.Orientation
+		timedIMUReadingResponse                     *TimedIMUReadingResponse
+		timedOdometerReadingResponse                *TimedOdometerReadingResponse
+		err                                         error
+	)
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, timedMovementSensorReadingTimeout)
+	defer cancel()
```

Comment:
I'm worried about how this section will interact with integration tests. We currently get data from the imu and lidar and compare the times, whichever one one occurs earlier we add in before grabbing new data for the one added. What will happen if imu is before lidar but odom data is after? We may need a way call each one individually.

Reply:
Oh interesting. I think either we could allow getting IMU data separately from odometry (as you said) or get all of the values together and give them the same timestamp.

Reply:
This problem will be resolved by changing the way we grab & insert data in offline mode.